### PR TITLE
Refactor LocalData (EXPOSUREAPP-2850)

### PR DESCRIPTION
**Pull request by jurajkusnier**
- Move encrypted shared preferences from `LocalData` object to other classes

### Testing
- Exhaustive Regression testing necessary
- Test upgrade from the older version
- Check if files are removed from the device
`databases/coronawarnapp-db`
`shared_preferences/shared_preferences_cwa.xml`

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/SecurityHelper.kt [hunk 1]
```diff
@@ -1,55 +1,36 @@
...
-import timber.log.Timber
 
 /**
  * Key Store and Password Access
  */
 object SecurityHelper {
 
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    internal val encryptedPreferencesProvider: (ApplicationComponent) -> SharedPreferences = {
+    internal val encryptedPreferencesProvider: (ApplicationComponent) -> SharedPreferences? = {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/SecurityHelper.kt:16]: Please make the access & retry mechanism part of `EncryptedPreferencesHelper`, I think the whole `object SecurityHelper` can then be deleted`.  I think `EncryptionErrorResetTool` can then also be almost deleted, `EncryptedPreferencesFactory` should then likely also be simplified and made part of `EncryptedPreferencesHelper`.

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorkerTest.kt [hunk 2]
```diff
@@ -126,19 +133,24 @@ class DiagnosisTestResultRetrievalPeriodicWorkerTest : BaseTest() {
...
             coVerify { submissionService.asyncRequestTestResult(registrationToken) }
-            coVerify { LocalData.isTestResultAvailableNotificationSent(true) }
             coVerify { testResultAvailableNotificationService.showTestResultAvailableNotification(testResult) }
             coVerify {
                 notificationHelper.cancelCurrentNotification(
                     NotificationConstants.NEW_MESSAGE_RISK_LEVEL_SCORE_NOTIFICATION_ID
                 )
             }
             assert(result is ListenableWorker.Result.Success)
+            assert(isTestResultAvailableNotificationSent.captured)
```
Comment (d4rken) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorkerTest.kt:143]: Please use kotest, i.e. `X shouldBe Y`

**Review by d4rken — CHANGES_REQUESTED** [review]
Code looks good, still need to test on device. Few changes I'd like to discuss, see comments.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/preferences/EncryptedPreferencesMigration.kt [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            Timber.d("EncryptedPreferencesMigration removing database $file")
+            SQLiteDatabase.deleteDatabase(file)
+        }
+    }
+
+    companion object {
+        private const val TRACING_DATABASE_NAME = "coronawarnapp-db"
+    }
+
+    // TODO: delete or update LocalData, SecurityConstants, SecurityHelper after all methods are migrated
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/preferences/EncryptedPreferencesMigration.kt]: removed

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/CoronaWarnApplication.kt [hunk 4]
```diff
@@ -70,10 +76,14 @@ class CoronaWarnApplication : Application(), HasAndroidInjector {
 
         CWADebug.initAfterInjection(component)
 
+        BackgroundWorkScheduler.init(component)
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/CoronaWarnApplication.kt:81]: yes

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/tracing/DefaultTracingStatus.kt [hunk 5]
```diff
@@ -72,14 +73,15 @@ class DefaultTracingStatus @Inject constructor(
         client.start()
             .addOnSuccessListener { cont.resume(it) }
             .addOnFailureListener { cont.resumeWithException(it) }
+            .also {
+                tracingSettings.isConsentGiven = true
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/nearby/modules/tracing/DefaultTracingStatus.kt:77]: It was there all the time. It was called `initialTracingActivationTimestamp()` before, but as we always used it only as a boolean value in `TracingPermissionHelper.isConsentGiven()` I rename it.
It was removed by mistake in previous PR https://github.com/corona-warn-app/cwa-app-android/commit/1b61142ff80d6e472b25b9f32371e610b2353b70#diff-429f8ee04350d96dcf9c0b0176310a0cf1db823097983b0ae865463833b267dc
So you can't see the proper diff here.

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorkerTest.kt (hunk 2 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/worker/DiagnosisTestResultRetrievalPeriodicWorkerTest.kt]: sure

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/preferences/PreferencesModule.kt [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package de.rki.coronawarnapp.storage.preferences
+
+import android.content.Context
+import android.content.pm.ApplicationInfo
+import dagger.Module
+import dagger.Provides
+import de.rki.coronawarnapp.util.di.AppContext
+
+@Module
+class PreferencesModule {
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/preferences/PreferencesModule.kt]: it was used only in preferences, but yes, it makes sense to have it in `AndroidModule`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/preferences/EncryptedPreferencesMigration.kt [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            Timber.e(e, "Database removing was not successful")
+        }
+    }
+
+    private fun copyData() {
+        if (!encryptedPreferencesHelper.isAvailable()) return
+
+        val encryptedSharedPreferences = SecurityHelper.globalEncryptedSharedPreferencesInstance
+        Timber.d("EncryptedPreferencesMigration START")
+        if (encryptedSharedPreferences != null) {
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/preferences/EncryptedPreferencesMigration.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/preferences/EncryptedPreferencesMigration.kt [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        fun wasInteroperabilityShown() = sharedPreferences.getBoolean(PREFERENCE_INTEROPERABILITY_WAS_USED, false)
+
+        fun isNotificationsRiskEnabled(): Boolean = sharedPreferences.getBoolean(PKEY_NOTIFICATIONS_RISK_ENABLED, true)
+
+        fun isNotificationsTestEnabled(): Boolean = sharedPreferences.getBoolean(PKEY_NOTIFICATIONS_TEST_ENABLED, true)
+
+        fun numberOfRemainingSharePositiveTestResultReminders(): Int =
+            sharedPreferences.getInt(PKEY_POSITIVE_TEST_RESULT_REMINDER_COUNT, Int.MIN_VALUE)
+
+        companion object {
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/preferences/EncryptedPreferencesMigration.kt:101]: no, updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/preferences/EncryptedPreferencesMigration.kt [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+                submissionSettings.registrationToken.update {
+                    registrationToken()
+                }
+                submissionSettings.initialTestResultReceivedAt = initialTestResultReceivedTimestamp().toInstantOrNull()
+                submissionSettings.devicePairingSuccessfulAt = devicePairingSuccessfulTimestamp().toInstantOrNull()
+                submissionSettings.isSubmissionSuccessful = numberOfSuccessfulSubmissions() >= 1
+                submissionSettings.isAllowedToSubmitKeys = isAllowedToSubmitDiagnosisKeys()
+            }
+
+            encryptedPreferencesHelper.clean()
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/preferences/EncryptedPreferencesMigration.kt]: The flow was updated. `isAvailable()` was removed and 'clean()` will be called always even if the migration fails

**Comment by harambasicluka** [comment]
Did on device testing: OnePlus 8 (Android 11) with a clean install, looks good. Will know also test an update scenario.

**Comment by harambasicluka** [comment]
@jurajkusnier test on an OnePlus 8 (Android 11). The app closed (no crash) after scanning a pending test (WRU-XD), will test the rest of the submission flow go on to an update scenario. I send you the logs of the corresponding device via email to not expose any internal urls.

**Comment by harambasicluka** [comment]
I have this flickering in the submission flow: teletan & qr code.
In the video you can see it twice in the end. I'll send you the logs via email.

https://user-images.githubusercontent.com/64483219/110805584-834a2a80-8281-11eb-9cd0-631d1acaf822.mp4

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/SecurityHelper.kt (hunk 1 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/SecurityHelper.kt]: I moved `SecurityHelper` to `EncryptedPreferencesHelper` and `EncryptionErrorResetTool` was updated as well

**Comment by harambasicluka** [comment]
Tested an update scenario on a Xiamoi MI9 (Android 9, WRU-XD), had a tested registered and changed some default settings and everything was like it was before the update 👍  But I also have here the flickering, will now test `release/1.15.x` to check if this PR is the cause.

**Comment by harambasicluka** [comment]
The flickering is already in `release/1.15.x` but not in `release/1.14.x`
https://github.com/corona-warn-app/cwa-app-android/pull/2569#issuecomment-796792376

**Review by d4rken — APPROVED** [review]
We will follow up with another PR tomorrow.

**Review by harambasicluka — APPROVED** [review]

**Review by BMItr — APPROVED** [review]

**Pull request merged**